File: Java/Calculator/src/main/java/com/example/calculator/Calculator.java
package com.example.calculator;

public class Calculator {

    public double add(double num1, double num2) {
        return num1 + num2;
    }

    public double subtract(double num1, double num2) {
        return num1 - num2;
    }

    public double multiply(double num1, double num2) {
        return num1 * num2;
    }

    public double divide(double num1, double num2) {
        if (num2 == 0) {
            throw new ArithmeticException("division by zero");
        }
        return num1 / num2;
    }

    public double power(double num1, double num2) {
        return Math.pow(num1, num2);
    }

    public double squareRoot(double num) {
        if (num < 0) {
            throw new ArithmeticException("cannot calculate square root of a negative number");
        }
        return Math.sqrt(num);
    }

    public int factorial(int num) {
        if (num < 0) {
            throw new IllegalArgumentException("cannot calculate factorial of a negative number");
        }
        if (num == 0) {
            return 1;
        }
        return num * factorial(num - 1);
    }

    public double logBase10(double num) {
        if (num <= 0) {
            throw new ArithmeticException("cannot calculate logarithm of a non-positive number");
        }
        return Math.log10(num);
    }

    public double naturalLog(double num) {
        if (num <= 0) {
            throw new ArithmeticException("cannot calculate natural logarithm of a non-positive number");
        }
        return Math.log(num);
    }

    public double sin(double num) {
        return Math.sin(num);
    }

    public double cos(double num) {
        return Math.cos(Math.toRadians(num));
    }

    public double tan(double num) {
        return Math.tan(Math.toRadians(num));
    }

    public String validateInputs(String num1, String num2) {
        try {
            Double.parseDouble(num1);
            Double.parseDouble(num2);
        } catch (NumberFormatException e) {
            return "input validation error";
        }
        return "";
    }
}

File: Python/Calculator/Calculator.py
import math


class Calculator:
    def __init__(self):
        self.reset()

    def reset(self):
        self.current_result = 0
        self.operands = []

    def input_number(self, num):
        self.operands.append(num)

    def add(self):
        self.current_result += sum(self.operands)
        self.current_result = round(self.current_result, 2)
        self.operands = []

    def subtract(self):
        self.current_result = self.operands[0]
        for operand in self.operands[1:]:
            self.current_result -= operand
        self.operands = []


    def multiply(self):
        self.current_result = math.prod(self.operands)
        self.operands = []

    def divide(self):
        if 0 in self.operands:
            raise ZeroDivisionError("Division by zero is not allowed")
        self.current_result /= math.prod(self.operands)
        self.operands = []

    def power(self):
        self.current_result = self.operands[0] ** self.operands[1]
        self.operands = []

    def square_root(self):
        if self.operands[0] < 0:
            raise ValueError("Square root of negative number is not allowed")
        self.current_result = math.sqrt(self.operands[0])
        self.operands = []

    def factorial(self):
        if self.operands[0] < 0:
            raise ValueError("Factorial of negative number is not allowed")
        self.current_result = math.factorial(int(self.operands[0]))
        self.operands = []

    def natural_log(self):
        if self.operands[0] <= 0:
            raise ValueError("Natural logarithm of non-positive number is not allowed")
        self.current_result = math.log(self.operands[0])
        self.operands = []

    def log_base_10(self):
        if self.operands[0] <= 0:
            raise ValueError("Logarithm of non-positive number is not allowed")
        self.current_result = math.log10(self.operands[0])
        self.operands = []

    def sine(self, in_radians=True):
        self.current_result = math.sin(self.operands[0] if in_radians else math.radians(self.operands[0]))
        self.operands = []

    def cosine(self, in_radians=True):
        self.current_result = math.cos(self.operands[0] if in_radians else math.radians(self.operands[0]))
        self.operands = []

    def tangent(self, in_radians=True):
        self.current_result = math.tan(self.operands[0] if in_radians else math.radians(self.operands[0]))
        self.operands = []

    def memory_store(self):
        self.memory = self.current_result

    def memory_recall(self):
        self.current_result = self.memory

    def memory_add(self):
        self.memory += self.current_result

    def memory_subtract(self):
        self.memory -= self.current_result

    def memory_clear(self):
        self.memory = 0

    def get_output(self):
        return self.current_result
